girder_worker/docker/tasks/__init__.py

```diff
@@ -0,0 +1,304 @@
+import sys
+import shutil
+import os
+try:
+    import docker
+    from docker.errors import DockerException
+    from girder_worker.docker import nvidia
+    from requests.exceptions import ReadTimeout
+except ImportError:
+    # These imports will not be available on the girder side.
+    pass
+from girder_worker.app import app, Task
+from girder_worker import logger
+from girder_worker.docker import utils
+from girder_worker.docker.stream_adapter import DockerStreamPushAdapter
+from girder_worker.docker.io import (
+    WriteStreamConnector,
+    ReadStreamConnector,
+    FileDescriptorReader,
+    StreamConnector,
+    StdStreamWriter
+)
+from girder_worker.docker.transforms import (
+    ContainerStdErr,
+    ContainerStdOut,
+    _TemporaryVolumeBase,
+    TemporaryVolume
+)
+from girder_worker_utils import _walk_obj
+
+
+BLACKLISTED_DOCKER_RUN_ARGS = ['tty', 'detach']
+
+
+def _pull_image(image):
+    """
+    Pulls the specified Docker image onto this worker.
+    """
+    client = docker.from_env(version='auto')
+    try:
+        client.images.pull(image)
+    except DockerException as dex:
+        logger.error('Error pulling Docker image %s:' % image)
+        logger.exception(dex)
+        raise
+
+
+def _run_container(image, container_args,  **kwargs):
+    # TODO we could allow configuration of non default socket
+    client = docker.from_env(version='auto')
+    if nvidia.is_nvidia_image(client.api, image):
+        client = nvidia.NvidiaDockerClient.from_env(version='auto')
+
+    logger.info('Running container: image: %s args: %s kwargs: %s'
+                % (image, container_args, kwargs))
+    try:
+        return client.containers.run(image, container_args, **kwargs)
+    except DockerException as dex:
+        logger.error(dex)
+        raise
+
+
+def _run_select_loop(task, container, read_stream_connectors, write_stream_connectors):
+    stdout = None
+    stderr = None
+    try:
+        # attach to standard streams
+        stdout = container.attach_socket(params={
+            'stdout': True,
+            'logs': True,
+            'stream': True
+        })
+
+        stderr = container.attach_socket(params={
+            'stderr': True,
+            'logs': True,
+            'stream': True
+        })
+
+        def exit_condition():
+            container.reload()
+            return container.status in ['exited', 'dead'] or task.canceled
```

Comment:
Silly nit, let's use `in {'exited', 'dead'}`, since what we really want is to detect presence in a set.

Reply:
Done